.metadata is an empty message even for missing metadata

In https://github.com/python/importlib_metadata/issues/489#issuecomment-2195177710, I learned that `Distribution.metadata` will return an empty `PackageMetadata` even when there is no metadata file present.

Although not part of the recommended operating practice, it seems to be occasional occurrence for an uninstalled package to leave a lingering metadata directory (`*.dist-info`) with nothing in it. Instead, I would expect `dist.metadata` to be `None` when no metadata file is present, and to return an empty `PackageMetadata` object when a metadata file is present but empty.

It seems the reason this expectation was previously missed is because (a) this case hasn't previously been tested, (b) the typespec indicates that [`email.message_from_string` accepts only a `str`](https://github.com/python/typeshed/blob/db7dd4cd54681291f2c7524aca8fdb1d59830e24/stdlib/email/__init__.pyi#L11), and (c) the implementation [suppresses the type error](https://github.com/python/importlib_metadata/blob/f3901686abc47853523f3b211873fc2b9e0c5ab5/importlib_metadata/__init__.py#L454) (introduced in d3908983078a47acecf430115b2bcc2d104f54a6).

The underlying reason `email.message_from_string` accepts None is because `io.StringIO` accepts `None` and is indistinguishable from `io.StringIO('')` for reading.

In 597776e8e4, I started exploring an optional None return type from `.metadata()`, but that does violate other interfaces (just internally):

```
_________________________________________________________ importlib_metadata/__init__.py _________________________________________________________
465: error: Value of type "PackageMetadata | None" is not indexable  [index]
475: error: Value of type "PackageMetadata | None" is not indexable  [index]
972: error: Incompatible return value type (got "PackageMetadata | None", expected "PackageMetadata")  [return-value]
1041: error: Value of type "PackageMetadata | None" is not indexable  [index]
_______________________________________________________ importlib_metadata/compat/py39.py ________________________________________________________
23: error: Value of type "PackageMetadata | None" is not indexable  [index]
```

--- a/importlib_metadata/__init__.py
+++ b/importlib_metadata/__init__.py
@@ -511,7 +511,7 @@ class Distribution(metaclass=abc.ABCMeta):
         return filter(None, declared)
 
     @property
-    def metadata(self) -> _meta.PackageMetadata:
+    def metadata(self) -> _meta.PackageMetadata | None:
         """Return the parsed metadata for this Distribution.
 
         The returned object will have keys that name the various bits of
@@ -521,10 +521,8 @@ class Distribution(metaclass=abc.ABCMeta):
         Custom providers may provide the METADATA file or override this
         property.
         """
-        # deferred for performance (python/cpython#109829)
-        from . import _adapters
 
-        opt_text = (
+        text = (
             self.read_text('METADATA')
             or self.read_text('PKG-INFO')
             # This last clause is here to support old egg-info files.  Its
@@ -532,7 +530,14 @@ class Distribution(metaclass=abc.ABCMeta):
             # (which points to the egg-info file) attribute unchanged.
             or self.read_text('')
         )
-        text = cast(str, opt_text)
+        return self._assemble_message(text)
+
+    @staticmethod
+    @pass_none
+    def _assemble_message(text: str) -> _meta.PackageMetadata:
+        # deferred for performance (python/cpython#109829)
+        from . import _adapters
+
         return _adapters.Message(email.message_from_string(text))
 
     @property

--- a/importlib_metadata/__init__.py
+++ b/importlib_metadata/__init__.py
@@ -543,7 +543,7 @@ class Distribution(metaclass=abc.ABCMeta):
     @property
     def name(self) -> str:
         """Return the 'Name' metadata for the distribution package."""
-        return self.metadata['Name']
+        return cast(PackageMetadata, self.metadata)['Name']
 
     @property
     def _normalized_name(self):
@@ -553,7 +553,7 @@ class Distribution(metaclass=abc.ABCMeta):
     @property
     def version(self) -> str:
         """Return the 'Version' metadata for the distribution package."""
-        return self.metadata['Version']
+        return cast(PackageMetadata, self.metadata)['Version']
 
     @property
     def entry_points(self) -> EntryPoints:
@@ -1050,7 +1050,7 @@ def distributions(**kwargs) -> Iterable[Distribution]:
     return Distribution.discover(**kwargs)
 
 
-def metadata(distribution_name: str) -> _meta.PackageMetadata:
+def metadata(distribution_name: str) -> _meta.PackageMetadata | None:
     """Get the metadata for the named package.
 
     :param distribution_name: The name of the distribution package to query.
@@ -1125,7 +1125,7 @@ def packages_distributions() -> Mapping[str, list[str]]:
     pkg_to_dist = collections.defaultdict(list)
     for dist in distributions():
         for pkg in _top_level_declared(dist) or _top_level_inferred(dist):
-            pkg_to_dist[pkg].append(dist.metadata['Name'])
+            pkg_to_dist[pkg].append(cast(PackageMetadata, dist.metadata)['Name'])
     return dict(pkg_to_dist)
 
 

--- a/importlib_metadata/__init__.py
+++ b/importlib_metadata/__init__.py
@@ -27,7 +27,7 @@ from contextlib import suppress
 from importlib import import_module
 from importlib.abc import MetaPathFinder
 from itertools import starmap
-from typing import Any, cast
+from typing import Any
 
 from . import _meta
 from ._collections import FreezableDefaultDict, Pair
@@ -38,6 +38,7 @@ from ._compat import (
 from ._functools import method_cache, pass_none
 from ._itertools import always_iterable, bucket, unique_everseen
 from ._meta import PackageMetadata, SimplePath
+from ._typing import md_none
 from .compat import py39, py311
 
 __all__ = [
@@ -543,7 +544,7 @@ class Distribution(metaclass=abc.ABCMeta):
     @property
     def name(self) -> str:
         """Return the 'Name' metadata for the distribution package."""
-        return cast(PackageMetadata, self.metadata)['Name']
+        return md_none(self.metadata)['Name']
 
     @property
     def _normalized_name(self):
@@ -553,7 +554,7 @@ class Distribution(metaclass=abc.ABCMeta):
     @property
     def version(self) -> str:
         """Return the 'Version' metadata for the distribution package."""
-        return cast(PackageMetadata, self.metadata)['Version']
+        return md_none(self.metadata)['Version']
 
     @property
     def entry_points(self) -> EntryPoints:
@@ -1125,7 +1126,7 @@ def packages_distributions() -> Mapping[str, list[str]]:
     pkg_to_dist = collections.defaultdict(list)
     for dist in distributions():
         for pkg in _top_level_declared(dist) or _top_level_inferred(dist):
-            pkg_to_dist[pkg].append(cast(PackageMetadata, dist.metadata)['Name'])
+            pkg_to_dist[pkg].append(md_none(dist.metadata)['Name'])
     return dict(pkg_to_dist)
 
 

--- a/importlib_metadata/_typing.py
+++ b/importlib_metadata/_typing.py
@@ -1,0 +1,15 @@
+import functools
+import typing
+
+from ._meta import PackageMetadata
+
+md_none = functools.partial(typing.cast, PackageMetadata)
+"""
+Suppress type errors for optional metadata.
+
+Although Distribution.metadata can return None when metadata is corrupt
+and thus None, allow callers to assume it's not None and crash if
+that's the case.
+
+# python/importlib_metadata#493
+"""

--- a/importlib_metadata/compat/py39.py
+++ b/importlib_metadata/compat/py39.py
@@ -4,7 +4,7 @@ Compatibility layer with Python 3.8/3.9
 
 from __future__ import annotations
 
-from typing import TYPE_CHECKING, Any, cast
+from typing import TYPE_CHECKING, Any
 
 if TYPE_CHECKING:  # pragma: no cover
     # Prevent circular imports on runtime.
@@ -12,7 +12,7 @@ if TYPE_CHECKING:  # pragma: no cover
 else:
     Distribution = EntryPoint = Any
 
-from .._meta import PackageMetadata
+from .._typing import md_none
 
 
 def normalized_name(dist: Distribution) -> str | None:
@@ -25,7 +25,7 @@ def normalized_name(dist: Distribution) -> str | None:
         from .. import Prepared  # -> delay to prevent circular imports.
 
         return Prepared.normalize(
-            getattr(dist, "name", None) or cast(PackageMetadata, dist.metadata)['Name']
+            getattr(dist, "name", None) or md_none(dist.metadata)['Name']
         )
 
 

--- a/importlib_metadata/compat/py39.py
+++ b/importlib_metadata/compat/py39.py
@@ -4,13 +4,15 @@ Compatibility layer with Python 3.8/3.9
 
 from __future__ import annotations
 
-from typing import TYPE_CHECKING, Any
+from typing import TYPE_CHECKING, Any, cast
 
 if TYPE_CHECKING:  # pragma: no cover
     # Prevent circular imports on runtime.
     from .. import Distribution, EntryPoint
 else:
     Distribution = EntryPoint = Any
+
+from .._meta import PackageMetadata
 
 
 def normalized_name(dist: Distribution) -> str | None:
@@ -22,7 +24,9 @@ def normalized_name(dist: Distribution) -> str | None:
     except AttributeError:
         from .. import Prepared  # -> delay to prevent circular imports.
 
-        return Prepared.normalize(getattr(dist, "name", None) or dist.metadata['Name'])
+        return Prepared.normalize(
+            getattr(dist, "name", None) or cast(PackageMetadata, dist.metadata)['Name']
+        )
 
 
 def ep_matches(ep: EntryPoint, **params) -> bool:

--- a/newsfragments/493.feature.rst
+++ b/newsfragments/493.feature.rst
@@ -1,0 +1,1 @@
+``.metadata()`` (and ``Distribution.metadata``) can now return ``None`` if the metadata directory exists but not metadata file is present.

--- a/tests/test_main.py
+++ b/tests/test_main.py
@@ -155,6 +155,16 @@ class InvalidMetadataTests(fixtures.OnSysPath, fixtures.SiteDir, unittest.TestCa
         dist = Distribution.from_name('foo')
         assert dist.version == "1.0"
 
+    def test_missing_metadata(self):
+        """
+        Dists with a missing metadata file should return None.
+
+        Ref python/importlib_metadata#493.
+        """
+        fixtures.build_files(self.make_pkg('foo-4.3', files={}), self.site_dir)
+        assert Distribution.from_name('foo').metadata is None
+        assert metadata('foo') is None
+
 
 class NonASCIITests(fixtures.OnSysPath, fixtures.SiteDir, unittest.TestCase):
     @staticmethod

